src/addrman.cpp

```diff
@@ -331,43 +329,68 @@ void CAddrMan::Attempt_(const CService& addr, int64_t nTime)
     info.nAttempts++;
 }
 
-CAddrInfo CAddrMan::Select_()
+CAddrInfo CAddrMan::Select_(bool newOnly)
 {
     if (size() == 0)
         return CAddrInfo();
 
+    // Track number of attempts to find a table entry, before giving up to avoid infinite loop
+    const int kMaxRetries = 200000;         // magic number so unit tests can pass
+    const int kRetriesBetweenSleep = 1000;
+    const int kRetrySleepInterval = 100;    // milliseconds
```

Comment:
Correct.  In normal operation, every 20s we exit to the outer loop which will try to get more addresses.  This is better than what happens now, which is an infinite loop with no exit possible.

The tests don't care how long it takes to find an address.  Since the tests are deterministic - each test will require the same number of loops - the sleep adds just over 3 mins to ./src/test/test_bitcoin.  If you want, I can remove the sleep when running in tests (by checking the instance variable nKey which is set to null by the tests).